### PR TITLE
Feature/system tests

### LINKED ISSUE #264: System Tests

### LINKED ISSUE #266: Fixing small parser issues for system tests

### LINKED ISSUE #269: backwards jump (jal) generates "Branch offset would invalidate program counter"
add x0, x0, x0
jal x0, -2

Only in 64Bit mode

### PR DESCRIPTION
**Pull request by ErikP0**
closes #264 fixes #269 closes #266 
This PR will add some tests for parsing and executing a program (testing parser, core and arch implementations).
The bugs that are discovered, should be fixed in this branch to see the progress.
Currently those bugs are left:
- Memory Reservation Directives don't recognize labels and constants
- `jalr` causes invalid program counter address in *super_sum*
- *factorial_rec* does not execute/ write assembled program into memory

### CODE REVIEW

File: tests/system/program-execution-test.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+using SuperSumTest64 = SuperSumTest<int64_t>;
+TEST_F(SuperSumTest64, supersum64) { superSumTest(); }
+
+template <typename SizeType>
+struct EndlessTest : public ProgramExecutionFixture<SizeType> {
+  EndlessTest() : ProgramExecutionFixture<SizeType>("endless.txt") {}
+
+  void endlessTest() {
+    this->assertForAllRegisters(
+        [](auto reg, auto val) { ASSERT_EQ(0, val) << "In Register " << reg; });
```
Comment (goldsborough) [inline tests/system/program-execution-test.cpp:241]: Please put `this` in the capture, we've had issues with GCC before

File: tests/system/program-execution-test.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+using SuperSumTest64 = SuperSumTest<int64_t>;
+TEST_F(SuperSumTest64, supersum64) { superSumTest(); }
+
+template <typename SizeType>
+struct EndlessTest : public ProgramExecutionFixture<SizeType> {
+  EndlessTest() : ProgramExecutionFixture<SizeType>("endless.txt") {}
+
+  void endlessTest() {
+    this->assertForAllRegisters(
```
Comment (goldsborough) [inline tests/system/program-execution-test.cpp]: No need for `this->` here. If anything add `super = ProgramExecutionFixture<SizeType>` and write `super::assertForAllRegisters` to indicate that this method is declared by the base class

File: tests/system/program-execution-test.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+using MemoryIOTest64 = MemoryIOTest<int64_t>;
+TEST_F(MemoryIOTest64, memoryio64) { memoryioTest(); }
+
+template <typename SizeType>
+struct SuperSumTest : public ProgramExecutionFixture<SizeType> {
+  SuperSumTest() : ProgramExecutionFixture<SizeType>("super_sum.txt") {}
+
+  void superSumTest() {
+    this->assertForAllRegisters(
```
Comment (goldsborough) [inline tests/system/program-execution-test.cpp]: See below

**Review by goldsborough — APPROVED** [review]

File: include/arch/riscv/jump-instructions.hpp [hunk 4]
```diff
@@ -291,7 +291,7 @@ class JumpAndLinkRegisterInstructionNode
       MemoryAccess& memoryAccess) const override {
     auto base = super::template _getChildValue<UnsignedWord>(memoryAccess, 1);
     auto offset =
-        super::template _getChildValue<UnsignedWord>(memoryAccess, 2) + base;
+        super::template _getChildValue<SignedWord>(memoryAccess, 2);
```
Comment (goldsborough) [inline include/arch/riscv/jump-instructions.hpp:294]: 👍

File: tests/system/program-execution-test.cpp (hunk 1 above)
Comment (ErikP0) [reply tests/system/program-execution-test.cpp:241]: I don't use any member (functions), so no this needed

File: include/parser/independent/section-directive.hpp [hunk 5]
```diff
@@ -61,7 +60,7 @@ class SectionDirective : public IntermediateDirective {
      * \param tracker The section tracker so we know in which section to reserve
    * our data.
      */
-  virtual void allocateMemory(const PreprocessingImmutableArguments& immutable,
+  virtual void allocateMemory(const AllocateMemoryImmutableArguments& immutable,
```
Comment (janschopohl) [inline include/parser/independent/section-directive.hpp:64]: add a forward declaration of `AllocateMemoryImmutableArguments`

File: include/parser/independent/memory-reservation-directive.hpp [hunk 6]
```diff
@@ -80,7 +79,7 @@ class MemoryReservationDirective : public IntermediateDirective {
    * \param tracker The section tracker so we know in which section to reserve
    * our data.
    */
-  virtual void allocateMemory(const PreprocessingImmutableArguments& immutable,
+  virtual void allocateMemory(const AllocateMemoryImmutableArguments& immutable,
```
Comment (janschopohl) [inline include/parser/independent/memory-reservation-directive.hpp:83]: add a forward declaration of `AllocateMemoryImmutableArguments`

**Review by janschopohl — APPROVED** [review]
Looks good!
:+1:

**Pull request merged**